File: crates/breez-sdk/bindings/examples/cli/langs/swift/Sources/BreezCLI/Issuer.swift
import Foundation
import BreezSdkSpark
import BigNumber

// MARK: - Issuer command names (for REPL completion)

let issuerCommandNames: [String] = [
    "issuer token-balance",
    "issuer token-metadata",
    "issuer create-token",
    "issuer mint-token",
    "issuer burn-token",
    "issuer freeze-token",
    "issuer unfreeze-token",
]

// MARK: - Dispatch

func dispatchIssuerCommand(_ args: [String], tokenIssuer: TokenIssuer) async {
    if args.isEmpty || args[0] == "help" {
        printIssuerHelp()
        return
    }

    let subName = args[0]
    let subArgs = Array(args.dropFirst())

    do {
        switch subName {
        case "token-balance":
            try await handleTokenBalance(tokenIssuer, subArgs)
        case "token-metadata":
            try await handleTokenMetadata(tokenIssuer, subArgs)
        case "create-token":
            try await handleCreateToken(tokenIssuer, subArgs)
        case "mint-token":
            try await handleMintToken(tokenIssuer, subArgs)
        case "burn-token":
            try await handleBurnToken(tokenIssuer, subArgs)
        case "freeze-token":
            try await handleFreezeToken(tokenIssuer, subArgs)
        case "unfreeze-token":
            try await handleUnfreezeToken(tokenIssuer, subArgs)
        default:
            print("Unknown issuer subcommand: \(subName). Use 'issuer help' for available commands.")
        }
    } catch {
        print("Error: \(error)")
    }
}

// MARK: - Help

private func printIssuerHelp() {
    print("\nIssuer subcommands:")
    print("  issuer \("token-balance".padding(toLength: 30, withPad: " ", startingAt: 0))Get issuer token balance")
    print("  issuer \("token-metadata".padding(toLength: 30, withPad: " ", startingAt: 0))Get issuer token metadata")
    print("  issuer \("create-token".padding(toLength: 30, withPad: " ", startingAt: 0))Create a new issuer token")
    print("  issuer \("mint-token".padding(toLength: 30, withPad: " ", startingAt: 0))Mint supply of the issuer token")
    print("  issuer \("burn-token".padding(toLength: 30, withPad: " ", startingAt: 0))Burn supply of the issuer token")
    print("  issuer \("freeze-token".padding(toLength: 30, withPad: " ", startingAt: 0))Freeze tokens at an address")
    print("  issuer \("unfreeze-token".padding(toLength: 30, withPad: " ", startingAt: 0))Unfreeze tokens at an address")
    print()
}

// MARK: - Handlers

// --- token-balance ---

private func handleTokenBalance(_ issuer: TokenIssuer, _ args: [String]) async throws {
    let result = try await issuer.getIssuerTokenBalance()
    printValue(result)
}

// --- token-metadata ---

private func handleTokenMetadata(_ issuer: TokenIssuer, _ args: [String]) async throws {
    let result = try await issuer.getIssuerTokenMetadata()
    printValue(result)
}

// --- create-token ---

private func handleCreateToken(_ issuer: TokenIssuer, _ args: [String]) async throws {
    let fp = FlagParser(args)
    guard let name = fp.get("name"),
          let ticker = fp.get("ticker") else {
        print("Usage: issuer create-token --name <name> --ticker <ticker> [--decimals N] [--freezable] [--max-supply N]")
        return
    }

    let decimals = fp.get("decimals").flatMap { UInt32($0) } ?? 6
    let freezable = fp.has("freezable", "f")
    let maxSupply: BInt = fp.get("max-supply").flatMap { BInt($0) } ?? BInt(0)

    let result = try await issuer.createIssuerToken(request: CreateIssuerTokenRequest(
        name: name,
        ticker: ticker,
        decimals: decimals,
        isFreezable: freezable,
        maxSupply: maxSupply
    ))
    printValue(result)
}

// --- mint-token ---

private func handleMintToken(_ issuer: TokenIssuer, _ args: [String]) async throws {
    guard !args.isEmpty, let amount = BInt(args[0]) else {
        print("Usage: issuer mint-token <amount>")
        return
    }

    let result = try await issuer.mintIssuerToken(request: MintIssuerTokenRequest(
        amount: amount
    ))
    printValue(result)
}

// --- burn-token ---

private func handleBurnToken(_ issuer: TokenIssuer, _ args: [String]) async throws {
    guard !args.isEmpty, let amount = BInt(args[0]) else {
        print("Usage: issuer burn-token <amount>")
        return
    }

    let result = try await issuer.burnIssuerToken(request: BurnIssuerTokenRequest(
        amount: amount
    ))
    printValue(result)
}

// --- freeze-token ---

private func handleFreezeToken(_ issuer: TokenIssuer, _ args: [String]) async throws {
    guard !args.isEmpty else {
        print("Usage: issuer freeze-token <address>")
        return
    }

    let result = try await issuer.freezeIssuerToken(request: FreezeIssuerTokenRequest(
        address: args[0]
    ))
    printValue(result)
}

// --- unfreeze-token ---

private func handleUnfreezeToken(_ issuer: TokenIssuer, _ args: [String]) async throws {
    guard !args.isEmpty else {
        print("Usage: issuer unfreeze-token <address>")
        return
    }

    let result = try await issuer.unfreezeIssuerToken(request: UnfreezeIssuerTokenRequest(
        address: args[0]
    ))
    printValue(result)
}

File: crates/breez-sdk/bindings/examples/cli/langs/kotlin-multiplatform/src/main/kotlin/Issuer.kt
import breez_sdk_spark.*
import com.ionspin.kotlin.bignum.integer.BigInteger
import org.jline.reader.LineReader

/**
 * Represents a single issuer subcommand.
 */
data class IssuerCliCommand(
    val name: String,
    val description: String,
    val run: suspend (issuer: TokenIssuer, reader: LineReader, args: List<String>) -> Unit
)

/**
 * All issuer subcommand names.
 */
val ISSUER_COMMAND_NAMES = listOf(
    "token-balance",
    "token-metadata",
    "create-token",
    "mint-token",
    "burn-token",
    "freeze-token",
    "unfreeze-token",
)

/**
 * Builds the issuer command registry.
 */
fun buildIssuerRegistry(): Map<String, IssuerCliCommand> {
    return mapOf(
        "token-balance" to IssuerCliCommand("token-balance", "Get issuer token balance", ::handleTokenBalance),
        "token-metadata" to IssuerCliCommand("token-metadata", "Get issuer token metadata", ::handleTokenMetadata),
        "create-token" to IssuerCliCommand("create-token", "Create a new issuer token", ::handleCreateToken),
        "mint-token" to IssuerCliCommand("mint-token", "Mint supply of the issuer token", ::handleMintToken),
        "burn-token" to IssuerCliCommand("burn-token", "Burn supply of the issuer token", ::handleBurnToken),
        "freeze-token" to IssuerCliCommand("freeze-token", "Freeze tokens at an address", ::handleFreezeToken),
        "unfreeze-token" to IssuerCliCommand("unfreeze-token", "Unfreeze tokens at an address", ::handleUnfreezeToken),
    )
}

/**
 * Dispatches an issuer subcommand.
 */
suspend fun dispatchIssuerCommand(
    args: List<String>,
    issuer: TokenIssuer,
    registry: Map<String, IssuerCliCommand>,
    reader: LineReader,
) {
    if (args.isEmpty() || args[0] == "help") {
        println()
        println("Issuer subcommands:")
        registry.keys.sorted().forEach { name ->
            val cmd = registry[name]!!
            println("  issuer %-30s %s".format(name, cmd.description))
        }
        println()
        return
    }

    val subName = args[0]
    val subArgs = args.drop(1)

    val cmd = registry[subName]
    if (cmd == null) {
        println("Unknown issuer subcommand: $subName. Use 'issuer help' for available commands.")
        return
    }

    try {
        cmd.run(issuer, reader, subArgs)
    } catch (e: Exception) {
        println("Error: ${e.message}")
    }
}

// ---------------------------------------------------------------------------
// Issuer command handlers
// ---------------------------------------------------------------------------

// --- token-balance ---

suspend fun handleTokenBalance(issuer: TokenIssuer, reader: LineReader, args: List<String>) {
    val result = issuer.getIssuerTokenBalance()
    printValue(result)
}

// --- token-metadata ---

suspend fun handleTokenMetadata(issuer: TokenIssuer, reader: LineReader, args: List<String>) {
    val result = issuer.getIssuerTokenMetadata()
    printValue(result)
}

// --- create-token ---

suspend fun handleCreateToken(issuer: TokenIssuer, reader: LineReader, args: List<String>) {
    val fp = FlagParser(args)
    val name = fp.getString("name")
    val ticker = fp.getString("ticker")
    val decimals = fp.getUInt("decimals") ?: 6u
    val freezable = fp.hasFlag("freezable")
    val maxSupplyStr = fp.getString("max-supply")

    if (name == null || ticker == null) {
        println("Usage: issuer create-token --name <name> --ticker <ticker> [--decimals N] [--freezable] [--max-supply N]")
        return
    }

    val maxSupply = if (maxSupplyStr != null) {
        try {
            BigInteger.parseString(maxSupplyStr)
        } catch (e: Exception) {
            println("Invalid max-supply: $maxSupplyStr")
            return
        }
    } else {
        BigInteger.ZERO
    }

    val result = issuer.createIssuerToken(
        CreateIssuerTokenRequest(
            name = name,
            ticker = ticker,
            decimals = decimals,
            isFreezable = freezable,
            maxSupply = maxSupply,
        )
    )
    printValue(result)
}

// --- mint-token ---

suspend fun handleMintToken(issuer: TokenIssuer, reader: LineReader, args: List<String>) {
    if (args.isEmpty()) {
        println("Usage: issuer mint-token <amount>")
        return
    }

    val amount = try {
        BigInteger.parseString(args[0])
    } catch (e: Exception) {
        println("Invalid amount: ${args[0]}")
        return
    }

    val result = issuer.mintIssuerToken(MintIssuerTokenRequest(amount = amount))
    printValue(result)
}

// --- burn-token ---

suspend fun handleBurnToken(issuer: TokenIssuer, reader: LineReader, args: List<String>) {
    if (args.isEmpty()) {
        println("Usage: issuer burn-token <amount>")
        return
    }

    val amount = try {
        BigInteger.parseString(args[0])
    } catch (e: Exception) {
        println("Invalid amount: ${args[0]}")
        return
    }

    val result = issuer.burnIssuerToken(BurnIssuerTokenRequest(amount = amount))
    printValue(result)
}

// --- freeze-token ---

suspend fun handleFreezeToken(issuer: TokenIssuer, reader: LineReader, args: List<String>) {
    if (args.isEmpty()) {
        println("Usage: issuer freeze-token <address>")
        return
    }

    val result = issuer.freezeIssuerToken(FreezeIssuerTokenRequest(address = args[0]))
    printValue(result)
}

// --- unfreeze-token ---

suspend fun handleUnfreezeToken(issuer: TokenIssuer, reader: LineReader, args: List<String>) {
    if (args.isEmpty()) {
        println("Usage: issuer unfreeze-token <address>")
        return
    }

    val result = issuer.unfreezeIssuerToken(UnfreezeIssuerTokenRequest(address = args[0]))
    printValue(result)
}
